src/fetchServices.ts

```diff
@@ -0,0 +1,189 @@
+#!/usr/bin/env node
+
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {google, servicemanagement_v1} from 'googleapis';
+const spreadsheetId = '14JYzl_8W3HyD0c1jYDXQJA2sVWBZiMoxT8Sp57xGWvk';
+
+/**
+ * Make a call to the Google Cloud Service Management API's services.list endpoint
+ * @return an array of hostnames ('foo.googleapis.com') for all available services
+ */
+export async function getAllServiceNames(): Promise<string[]> {
+  const auth = await google.auth.getClient({
+    scopes: ['https://www.googleapis.com/auth/cloud-platform'],
+  });
+  const client = google.servicemanagement({
+    version: 'v1',
+    auth: auth,
+  });
+  const serviceNames: string[] = [];
+  return await listServices(client, serviceNames);
+}
+
+export async function listServices(
+  client: servicemanagement_v1.Servicemanagement,
+  serviceNames: string[],
+  nextPageToken = ''
+): Promise<string[]> {
+  const res = await client.services.list(
+    {
+      pageSize: 100,
+      pageToken: nextPageToken,
+    },
+    {
+      retryConfig: {
+        retry: 10,
+      },
+    }
+  );
+
+  if (res.data.services) {
+    res.data.services.map(item => {
+      if (item.serviceName) {
+        serviceNames.push(item.serviceName);
+      }
+    });
+  }
+
+  if (res.data.nextPageToken) {
+    await listServices(client, serviceNames, res.data.nextPageToken);
+  }
+  return serviceNames;
+}
+
+/**
+ * Make a call to the Google Cloud Service Management API's services.getConfig endpoint
+ * @param serviceName - the hostname of a service ('foo.googleapis.com')
+ * @return the service config of the passed hostname
+ */
+export async function getServiceConfig(
+  serviceName: string
+): Promise<servicemanagement_v1.Schema$Service> {
+  const auth = await google.auth.getClient({
```

Comment:
Ok - this appears to be called in a loop.  You really only want to do the auth step here a single time.  I would lift this, and the instantiation of the service management client to the outer body of the file.  Something like:

```js
const auth = new google.auth.GoogleAuth({ 
  scopes: <scopey stuff>
});
const servicemanagement = google.servicemanagement({
    version: 'v1',
    auth: auth,
});
```

Except - just do it once, in the outer scope of the file (not in a function).   Notice - I am using `new google.auth.GoogleAuth` up there instead of `await google.auth.getClient`.  That is important for this style.